module/zfs/zio.c

```diff
@@ -4431,7 +4431,7 @@ zio_ready(zio_t *zio)
 	}
 
 	if (zio->io_flags & ZIO_FLAG_NODATA) {
-		if (BP_IS_GANG(bp)) {
+		if (bp != NULL && BP_IS_GANG(bp)) {
```

Comment:
This check seems reasonable enough. I was just wondering do you happen to know when exactly `bp == NULL` might happen in this case though? I was trying to spot or understand why this might happen.

Reply:
I do not know of any case in the current code base that would trigger this. This bug should be dormant until something begins to use it.

Reply:
Would it be possible to just add in an `ASSERT3P(bp, !=, NULL)` to get rid of the CLANG static analyzer complaint instead of `if (bp != NULL`? I am just thinking we might to make sure that `bp != NULL` here in the future with the `ASSERT()`. The only reason I am suggesting this is because I am having a hard time even spotting when `ZIO_FLAG_NODATA` is set but there is no associated BP. I would think adding in the `ASSERT()` might be better as the assumption at this point is that is never `NULL` in this case.

Reply:
It is possible, but I do not like the idea of doing that here for a few reasons:

1. If we want to assert `zio->io_bp != NULL` in pipeline stages to prevent NULL pointer dereferences, there are many other places that are a higher priority than this one. For example, `zio_read_bp_init()`, `zio_write_bp_init()`, `zio_gang_assemble()`, `zio_dva_allocate()`, `zio_dva_free()`, `zio_dva_claim()` and possibly others will unconditionally have a NULL pointer dereference before `zio_ready()` is called. However, when we add an assertion, it should guard against a precondition that we cannot handle properly. In this case, there is no reason why we cannot handle a NULL bp in `zio_ready()`.

2. `zio_ready()` is part of the interlock pipeline with `zio_done()`. That is used as a subpipeline for almost every zio pipeline, with the only exception being `ZIO_VDEV_CHILD_PIPELINE`. `zio_done()` supports being given a NULL `zio->io_bp` under most conditions, so I would prefer not to add an artificial restriction on what can reach `zio_done()` in `zio_ready()`, even if we are not using things that way right now.

Reply:
@ryao maybe wrong here, but even though you sighted points that may fail on static analysis, I am still confused over how an `ASSERT` doesn't achieve the same concept. Feel free to rip me apart if I am wrong. As far as I know, the only true error code free code is filled with assets that catches the worse case. What I am missing with your tool that can handle dynamic code development that isn't handled by good `ASSERT` statements? My point is, if you want your tool to stop gripping, but an assumed thing is true, then we shouldn't replace it with a false check that we assume is true. What I am missing here?